x-pack/packages/kbn-elastic-assistant/impl/assistant/prompt_editor/system_prompt/system_prompt_modal/system_prompt_settings.tsx

```diff
@@ -128,9 +132,70 @@ export const SystemPromptSettings: React.FC<Props> = React.memo(
+          Object.values(conversationSettings).forEach((convo) => {
+            if (convo.id !== convo.title) {
+              updatedConversationsSettingsBulkActions = {
+                ...updatedConversationsSettingsBulkActions,
+                update: {
+                  ...(updatedConversationsSettingsBulkActions.update ?? {}),
+                  [convo.id]: {
+                    ...(updatedConversationsSettingsBulkActions.update
+                      ? updatedConversationsSettingsBulkActions.update[convo.id] ?? {}
+                      : {}),
+                    apiConfig: {
+                      ...((updatedConversationsSettingsBulkActions.update
+                        ? updatedConversationsSettingsBulkActions.update[convo.id] ?? {}
+                        : {}
+                      ).apiConfig ?? {}),
+                      defaultSystemPromptId: currentPromptConversationIds.includes(convo.id)
+                        ? selectedSystemPrompt?.id
+                        : convo.apiConfig.defaultSystemPromptId === selectedSystemPrompt?.id
+                        ? // remove the default System Prompt if it is assigned to a conversation
+                          // but that conversation is not in the currentPromptConversationList
+                          // This means conversation was removed in the current transaction
+                          undefined
+                        : //  leave it as it is .. if that conversation was neither added nor removed.
+                          convo.apiConfig.defaultSystemPromptId,
+                    },
+                  },
+                },
+              };
+            } else {
+              updatedConversationsSettingsBulkActions = {
+                ...updatedConversationsSettingsBulkActions,
+                create: {
+                  ...(updatedConversationsSettingsBulkActions.create ?? {}),
+                  [convo.id]: {
+                    ...convo,
+                    apiConfig: {
+                      ...convo.apiConfig,
+                      defaultSystemPromptId: currentPromptConversationIds.includes(convo.id)
+                        ? selectedSystemPrompt?.id
+                        : convo.apiConfig.defaultSystemPromptId === selectedSystemPrompt?.id
+                        ? // remove the default System Prompt if it is assigned to a conversation
+                          // but that conversation is not in the currentPromptConversationList
+                          // This means conversation was removed in the current transaction
+                          undefined
+                        : //  leave it as it is .. if that conversation was neither added nor removed.
+                          convo.apiConfig.defaultSystemPromptId,
+                    },
+                  },
+                },
+              };
+            }
+          });
+          setConversationsSettingsBulkActions(updatedConversationsSettingsBulkActions);
```

Comment:
Consider refactoring `136` - `189`  into several smaller, (and perhaps pure) functions, because the current implementation requires reasoning about multiple mutations of `updatedConversationsSettingsBulkActions`, and the following block:

```ts
                        ? selectedSystemPrompt?.id
                        : convo.apiConfig.defaultSystemPromptId === selectedSystemPrompt?.id
                        ? // remove the default System Prompt if it is assigned to a conversation
                          // but that conversation is not in the currentPromptConversationList
                          // This means conversation was removed in the current transaction
                          undefined
                        : //  leave it as it is .. if that conversation was neither added nor removed.
                          convo.apiConfig.defaultSystemPromptId,
```

appears to be nearly identical in the `if (convo.id !== convo.title)` / `else` branches